### PR TITLE
feat: Add actions to collect consumer latency metrics

### PR DESCRIPTION
**Pull request by bmcquilkin-sentry**
Adds ability to collect consumer latency to see if consumers are lagging behind or failing. Reports histogram metrics to Datadog which is also added as a dependency

### CODE REVIEW

**Review by fpacifici — COMMENTED** [review]
Please add unit tests for the action you added.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+def measure_cluster(
+    cluster_name: str, config: ClusterConfig, topics: list[str]
+) -> list[TopicConsumerLatency]:
+    consumer_group_id = f"consumer-latency-group-{cluster_name}"
+    consumer_client_id = f"consumer-latency-client-{cluster_name}"
+
+    client_config = create_kafka_client_config(config)
+
+    admin = AdminClient(client_config)
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: See the  comment by cursor, You do not need to reimplement the config to create an AdminClient. 
Please use this https://github.com/getsentry/sentry-kafka-management/blob/fd012cd7ca8df9ab641c9fad7eb5d5650fd7bf5b/sentry_kafka_management/connectors/admin.py#L8

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+@dataclass
+class TopicConsumerLatency:
+    cluster_name: str
+    topic_name: str
+    group_id: str
+    latency_ms: float
+    timestamp: int | None
+
+
+def create_kafka_client_config(config: ClusterConfig) -> dict[str, object]:
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: You do not need this method. Please see https://github.com/getsentry/sentry-kafka-management/pull/73/changes#r3256072857

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+            client_config["sasl.password"] = config["sasl_password"]
+        else:
+            client_config["sasl.password"] = os.path.expandvars(config["sasl_password"])
+
+    return client_config
+
+
+def list_consumer_group_ids(admin: AdminClient) -> list[str]:
+    """Get all consumer group IDs on the cluster."""
+    result = admin.list_consumer_groups(request_timeout=10.0).result()
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Why did you have to set the request timeout ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+    errors: list[KafkaException] = result.errors
+    valid: list[ConsumerGroupListing] = result.valid
+
+    for error in errors:
+        raise error
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply sentry_kafka_management/actions/latency/consumer_latency.py]: THis concern is valid. 
In order to raise without loosing information, create a new exception class (extend Exception) and make it contain a list of errors so that you can print them all when the message is printed out

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+    group_ids: list[str] = []
+
+    for listing in valid:
+        group_ids.append(listing.group_id)
+
+    return sorted(group_ids)
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: nit: 
```
return sorted([listing.group_id for listing in valid])
```

Also why do you need to sort the output ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+def get_committed_offsets(admin: AdminClient, group_id: str) -> list[TopicPartition]:
+    """Get the committed offsets for a consumer group."""
+    result: ConsumerGroupTopicPartitions = admin.list_consumer_group_offsets(
+        [ConsumerGroupTopicPartitions(group_id)]
+    )[group_id].result()
+
+    committed: list[TopicPartition] = []
+
+    for partition in result.topic_partitions:
+        if partition.error is None:
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: We should not hide the error.
Please raise the error if the result is not success.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+    topic_metadata = metadata.topics.get(topic)
+
+    if topic_metadata is None:
+        return []
+
+    if topic_metadata.error is not None:
+        return []
+
+    return sorted(topic_metadata.partitions)
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Not sure I get why you need to extract the topics and the partitions from the active consumer. 
`describe_consumer_groups` and `describe_topics`  on AdminClient already give you the number of partitions for a topic.
`get_committed_offsets` also would give you all partitions

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    deadline = time.monotonic() + 10.0
+
+    while time.monotonic() < deadline:
+        msg = consumer.poll(1.0)
+
+        if msg is None:
+            continue
+
+        if msg.error():
+            return None
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Please do not hide errors

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+    return sorted(topic_metadata.partitions)
+
+
+def read_timestamp_ms(consumer: Consumer, topic: str, partition: int, offset: int) -> int | None:
+    """Read the timestamp of a message from a Kafka topic."""
+    consumer.assign([TopicPartition(topic, partition, offset)])
+    deadline = time.monotonic() + 10.0
+
+    while time.monotonic() < deadline:
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py:111]: Pass the timeout as a parameter please

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+            return None
+
+        return int(ts_ms)
+
+    return None
+
+
+def get_partition_latency(
+    consumer: Consumer, topic: str, partition: int, committed_offset: int
+) -> tuple[float, int | None] | None:
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Why would this not be a `float` ? 
Specifically:
- Why do we need the ts out ?
- When would ts be None in a normal scenario ? 
- When would this return None ? Even if there was no data in the topic we should assume 0.0 latency.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+    """Get the latency of a partition."""
+
+    low, high = consumer.get_watermark_offsets(
+        TopicPartition(topic, partition), timeout=10.0, cached=False
+    )
+
+    if high == OFFSET_INVALID:
+        return None
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Shouldn't this be an error ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+
+    return None
+
+
+def get_partition_latency(
+    consumer: Consumer, topic: str, partition: int, committed_offset: int
+) -> tuple[float, int | None] | None:
+    """Get the latency of a partition."""
+
+    low, high = consumer.get_watermark_offsets(
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py:149]: What does this return in these cases ?
- the partition is empty ?
- either low or high are beyond retention ? 

I see you manage some corner cases below but I do not know which one is which.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+        TopicPartition(topic, partition), timeout=10.0, cached=False
+    )
+
+    if high == OFFSET_INVALID:
+        return None
+
+    if high <= low or committed_offset >= high:
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: why would high be < low ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+    if ts_ms is None:
+        return None
+
+    now_ms = int(time.time() * 1000)
+    latency_ms = float(now_ms - ts_ms)
+    return latency_ms, ts_ms
+
+
+def measure_cluster(
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: This is the main method to collect latency. Please rename to `get_cluster_latency`

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+
+    client_config = create_kafka_client_config(config)
+
+    admin = AdminClient(client_config)
+
+    consumer = Consumer(
+        {
+            "enable.auto.commit": False,
+            "group.id": consumer_group_id,
+            "client.id": consumer_client_id,
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: No need to set the client id.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+    consumer_client_id = f"consumer-latency-client-{cluster_name}"
+
+    client_config = create_kafka_client_config(config)
+
+    admin = AdminClient(client_config)
+
+    consumer = Consumer(
+        {
+            "enable.auto.commit": False,
+            "group.id": consumer_group_id,
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: You should not need to set the group id when assigning partitions to yourself. Actually assigning one is going to be persistent in kafka and we will start measuring the latency of this consumer group as well, which we do not want.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+            "client.id": consumer_client_id,
+            **client_config,
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Id you need a consumer config which looks like the one of AdminClient, please refactor https://github.com/getsentry/sentry-kafka-management/blob/fd012cd7ca8df9ab641c9fad7eb5d5650fd7bf5b/sentry_kafka_management/connectors/admin.py#L8.

You can have a kafka_config.py file in the connectors package that is used both by this function and by the admin.py file.
So that we do not duplicate this logic.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+
+            committed_offsets = {
+                (tp.topic, tp.partition): int(tp.offset)
+                for tp in get_committed_offsets(admin, group_id)
+            }
+            committed_topics = {topic for topic, _partition in committed_offsets}
+
+            for topic in topics:
+                if topic not in committed_topics:
+                    continue
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: when can this happen ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+
+                latency_ms = 0.0
+                oldest_outstanding_timestamp_ms: int | None = None
+                has_topic_latency = False
+
+                for partition in get_topic_partitions(consumer, topic):
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Don't you already have the partitions defined here ?
https://github.com/getsentry/sentry-kafka-management/pull/73/changes#diff-950f4cb987acff132701c84bec13ae2e9e8a843349a49255adcdd33ed74385baR185

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+                    latency_ms = max(latency_ms, partition_latency_ms)
+                    has_topic_latency = True
+                    if message_timestamp_ms is not None:
+                        if oldest_outstanding_timestamp_ms is None:
+                            oldest_outstanding_timestamp_ms = message_timestamp_ms
+                        else:
+                            oldest_outstanding_timestamp_ms = min(
+                                oldest_outstanding_timestamp_ms,
+                                message_timestamp_ms,
+                            )
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: I am not sure I follow what you need `oldest_outstanding_timestamp_ms` for.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+        consumer.close()
+
+    return scans
+
+
+def get_configured_topic_names(config: YamlKafkaConfig, cluster_name: str) -> list[str]:
+    try:
+        return list(config.get_topics_config(cluster_name))
+    except KeyError:
+        return []
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Why would we need to suppress the KeyError in case a bad cluster is passed ?

File: sentry_kafka_management/actions/latency/metrics.py [hunk 22]
```diff
@@ -0,0 +1,76 @@
+from __future__ import annotations
+
+from enum import Enum
+from typing import Protocol
+
+from datadog.dogstatsd.base import DogStatsd
+
+Tags = dict[str, str]
+
+METRICS_PREFIX = "kafka.broker"
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/metrics.py]: Please call this `kafka.topic`. kafka.broker are the metrics produced by kafka itself

File: sentry_kafka_management/actions/latency/metrics.py [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+        )
+        self.datadog_client.enable_background_sender(
+            sender_queue_size=SENDER_QUEUE_SIZE,
+            sender_queue_timeout=SENDER_QUEUE_TIMEOUT,
+        )
+        self.__tags = tags or {}
+
+    @staticmethod
+    def __normalize_tags(tags: Tags) -> list[str]:
+        return [f"{key}:{value.replace('|', '_')}" for key, value in tags.items()]
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/metrics.py]: You should not need to normalize tags. We know what they contain.

File: sentry_kafka_management/scripts/latency/consumer_latency.py [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+) -> None:
+    """
+    Emit Kafka consumer latency metrics for configured clusters.
+    """
+    kafka_config = YamlKafkaConfig(config)
+    metrics_backend = DatadogMetricsBackend(statsd_host, statsd_port)
+
+    while True:
+        run_latency_metrics_action(kafka_config, metrics_backend)
+        time.sleep(interval)
```
Comment (fpacifici) [inline sentry_kafka_management/scripts/latency/consumer_latency.py:85]: We need logs here to understand what the system is doing.
At every iteration, log which topics we go latency from (you can return them from run_latency_metrics_actions).

File: tests/scripts/test_latency.py
Comment (fpacifici) [inline tests/scripts/test_latency.py:1]: two high level comments on the tests.
1. Please focus on unit tests first for the consumer_latency module. Thsoe are going to be the highest value tests as you can cover corner cases.
2. the integration tests for the script can be few high level ones.
3. Try to avoid mocking everything you do not need to. 

```
@patch("sentry_kafka_management.scripts.latency.consumer_latency.time.sleep")
@patch("sentry_kafka_management.scripts.latency.consumer_latency.run_latency_metrics_action")
@patch("sentry_kafka_management.scripts.latency.consumer_latency.DatadogMetricsBackend")
@patch("sentry_kafka_management.scripts.latency.consumer_latency.YamlKafkaConfig")
```
You do not want to mock YamlKafkaConfig, can use the real one. 
When mocking sleep. please mock the `time` module not the version imported by the consumer_latency modules

**Review by fpacifici — CHANGES_REQUESTED** [review]
Please see the comments inline

File: sentry_kafka_management/actions/latency/consumer_latency.py (hunk 16 above)
Comment (bmcquilkin-sentry) [reply sentry_kafka_management/actions/latency/consumer_latency.py]: Confluent Kafka says that [group.id must be set](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Consumer.Consumer), the group is skipped over and doesn't have its latency calculated

Comment (fpacifici) [reply sentry_kafka_management/actions/latency/consumer_latency.py]: Alright, then could you please verify that, after you run your script, if you try to describe the consumer group you do not see any committed offset ? Otherwise we would get datadog metrics for that consumer group as well.

File: sentry_kafka_management/actions/latency/consumer_latency.py (hunk 12 above)
Comment (bmcquilkin-sentry) [reply sentry_kafka_management/actions/latency/consumer_latency.py:149]: Zero when the partition is empty and whatever oldest is available for the other case. I will add a comment to note this behavior

File: sentry_kafka_management/scripts/latency/consumer_latency.py [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+    "--statsd-port",
+    required=True,
+    type=int,
+    help="DogStatsD port to emit metrics to",
+)
+@click.option(
+    "-i",
+    "--interval",
+    required=False,
+    default=1.0,
```
Comment (fpacifici) [inline sentry_kafka_management/scripts/latency/consumer_latency.py]: Higher please. 10 seconds is a good time

File: sentry_kafka_management/scripts/latency/consumer_latency.py [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    Emit Kafka consumer latency metrics for configured clusters.
+    """
+    kafka_config = YamlKafkaConfig(config)
+    metrics_backend = DatadogMetricsBackend(statsd_host, statsd_port)
+
+    click.echo(f"Starting consumer latency collection (interval={interval:.3f}s)")
+
+    while True:
+        scans = run_latency_metrics_action(kafka_config, metrics_backend, timeout)
```
Comment (fpacifici) [inline sentry_kafka_management/scripts/latency/consumer_latency.py]: nit: naming. A good function name will tell you what this function does without you having to read the implementation. 
Now `run_latency_metrics_action` tells me it runs an action about latency metrics. What is it doing about latency metrics ?
What about `record_consumer_group_latency`  ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+    ts_ms = read_timestamp_ms(consumer, topic, partition, committed_offset, timeout)
+
+    now_ms = int(time.time() * 1000)
+    return float(now_ms - ts_ms)
+
+
+def get_cluster_latency(
+    cluster_name: str,
+    config: ClusterConfig,
+    topics: list[str],
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Why do we even need to pass topics here?
We are already getting them when describing the consumer group. 

Is there any reason to exclude topics ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+
+                    latency_ms = max(latency_ms, partition_latency_ms)
+
+                scans.append(
+                    TopicConsumerLatency(
+                        cluster_name=cluster_name,
+                        group_id=group_id,
+                        topic_name=topic,
+                        latency_ms=latency_ms,
+                    )
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py:236]: Feel free to record latency by partition. We do that for backlog and it is useful. Now we cannot do that in the existing system as we record latency in the consumer, but it is likely going to be useful to be able to break it down by partition.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 29]
```diff
@@ -0,0 +1,230 @@
...
+    result: ConsumerGroupTopicPartitions = admin.list_consumer_group_offsets(
+        [ConsumerGroupTopicPartitions(group_id)]
+    )[group_id].result()
+
+    committed: list[TopicPartition] = []
+
+    for partition in result.topic_partitions:
+        if partition.error is not None:
+            raise KafkaException(partition.error)
+        committed.append(partition)
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply sentry_kafka_management/actions/latency/consumer_latency.py:95]: This is not wrong. Let's add a TODO and a follow up ticket to manage topics where some partitions have commit and others have messages but no commit. If a consumer never commits a partition while the partition keeps receiving traffic we should record that latency.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 30]
```diff
@@ -0,0 +1,223 @@
...
+    timeout: int,
+) -> float:
+    """Get the latency of a partition."""
+
+    low, high = consumer.get_watermark_offsets(
+        TopicPartition(topic, partition), timeout=timeout, cached=False
+    )
+
+    if high == OFFSET_INVALID:
+        raise ValueError(f"No valid high watermark for {topic}[{partition}]")
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: When can this happen ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 31]
```diff
@@ -0,0 +1,223 @@
...
+
+    if high == low or committed_offset >= high:
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Can `high == low` happen when the partition has only one message and that message was sent hours ago ? In this case the consumer would not be caught up.

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 32]
```diff
@@ -0,0 +1,223 @@
...
+        return 0.0  # Empty partition or caught up
+
+    if committed_offset < 0:
+        return 0.0  # No committed offset
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: Can it actually be negative ?

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 33]
```diff
@@ -0,0 +1,223 @@
...
+    deadline = time.monotonic() + timeout
+
+    while time.monotonic() < deadline:
+        msg = consumer.poll(1.0)
+        if msg is None:
+            continue
+
+        error = msg.error()
+        if error is not None:
+            raise KafkaException(error)
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: There are some errors that we should retry https://github.com/getsentry/arroyo/blob/main/arroyo/backends/kafka/consumer.py#L161-L165

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 34]
```diff
@@ -0,0 +1,223 @@
...
+            **build_broker_config(config),
+        }
+    )
+
+    checked_topics = set(topics)
+    scans: list[TopicConsumerLatency] = []
+
+    try:
+        group_ids = list_consumer_group_ids(admin)
+        for group_id in group_ids:
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py:201]: I think we should continue till the end of the run before failing in case something wrong happen.
Imagine if we misconfigure a cluster, we want to still record the metrics from the good ones. Then we make the script fail, but, in the meantime, we still have good metrics.

File: tests/actions/latency/test_consumer_latency.py [hunk 35]
```diff
@@ -0,0 +1,455 @@
...
+    def __init__(self) -> None:
+        self.histograms: list[tuple[str, int | float, Tags | None]] = []
+
+    def histogram(self, name: str, value: int | float, tags: Tags | None = None) -> None:
+        self.histograms.append((name, value, tags))
+
+
+def _make_group_listing(group_id: str) -> Mock:
+    listing = Mock()
+    listing.group_id = group_id
```
Comment (fpacifici) [inline tests/actions/latency/test_consumer_latency.py]: Consider returning the actual data type the admin interface will return. That will properly test your function. Imagine  if it turned out the group_id did not exist on the returned object by the admin interface. The test would pass and the code would be broken.
This is a change that is quite easy to implement and raise the quality of your test. Removing the mock from the AdminClient is another matter, in that case the complexity is not worth the test quality improvement.

File: tests/actions/latency/test_consumer_latency.py [hunk 36]
```diff
@@ -0,0 +1,455 @@
...
+    bad_tp.topic = "topic-a"
+    bad_tp.partition = 0
+    bad_tp.offset = 100
+    bad_tp.error = KafkaError(KafkaError.UNKNOWN, "error")
+    admin.list_consumer_group_offsets.return_value = {
+        "group-a": _make_committed_offsets_future([bad_tp])
+    }
+
+    with pytest.raises(KafkaException):
+        get_committed_offsets(admin, "group-a")
```
Comment (fpacifici) [inline tests/actions/latency/test_consumer_latency.py]: These three tests are very similar, just different input and different output. You can use parametrize to make them more concise: https://docs.pytest.org/en/stable/example/parametrize.html#different-options-for-test-ids

**Review by fpacifici — COMMENTED** [review]
Better, a few more comments

File: tests/actions/latency/test_consumer_latency.py (hunk 35 above)
Comment (bmcquilkin-sentry) [reply tests/actions/latency/test_consumer_latency.py]: Noting that in order to construct `Message` objects we need to update to confluent-kafka 2.13 (we are currently on 2.8) and there seem to be some incompatibilities shown in mypy when upgrading

File: sentry_kafka_management/actions/latency/consumer_latency.py [hunk 37]
```diff
@@ -0,0 +1,268 @@
...
+    consumer: Consumer | None = None
+
+    retentions_by_topic: dict[str, int] = {}
+
+    for topic_name, topic_config in topics.items():
+        retention_ms = topic_config["settings"].get("retention.ms")
+        if retention_ms is None:
+            errors.append(
+                KeyError(f"Topic '{topic_name}' settings missing required 'retention.ms' field")
+            )
```
Comment (fpacifici) [inline sentry_kafka_management/actions/latency/consumer_latency.py]: That's fine. there is a sensible default. Assume the default is 7 days.

**Review by fpacifici — APPROVED** [review]
Please address the issue with the default retention before merging.
